quarkus-resteasy does not properly guard against invalid URLs

### Describe the bug

During an automated fuzzing session, we discovered that quarkus-resteasy doesn't deal well with malformed URLs. Basically, since `VertxUtil#extractUriInfo` is called very early on in request processing, its failure will lead to a 500 (and possibly a stacktrace) where it really should be a 400.
To make that function fail, you can use invalid URL encoding - e.g., do `/%FF`. You'll end up with a `java.nio.charset.MalformedInputException`.
This isn't a security-critical issue, but on the other hand it also shouldn't flood your logs with stacktraces.

### Expected behavior

The call should return HTTP 400, Bad Request.

### Actual behavior

Instead, we get a HTTP 500, possibly including a full stacktrace. For the reproducer below, you'll get
```
< HTTP/1.1 500 Internal Server Error
< content-type: application/json; charset=utf-8
< content-length: 1917
< 
* Connection #0 to host localhost left intact
{"details":"Error id a7604c2a-41c6-40f9-9f3f-8eae21129131-3, java.lang.RuntimeException: java.nio.charset.MalformedInputException: Input length = 1","stack":"java.lang.RuntimeException: java.nio.charset.MalformedInputException: Input length = 1\n\tat org.jboss.resteasy.util.Encode.decodeBytes(Encode.java:252)\n\tat org.jboss.resteasy.util.Encode.decodePath(Encode.java:228)\n\tat org.jboss.resteasy.specimpl.ResteasyUriInfo$InitData.<init>(ResteasyUriInfo.java:577)\n\tat org.jboss.resteasy.specimpl.ResteasyUriInfo.initialize(ResteasyUriInfo.java:73)\n\tat org.jboss.resteasy.specimpl.ResteasyUriInfo.<init>(ResteasyUriInfo.java:65)\n\tat org.jboss.resteasy.specimpl.ResteasyUriInfo.<init>(ResteasyUriInfo.java:56)\n\tat io.quarkus.resteasy.runtime.standalone.VertxUtil.extractUriInfo(VertxUtil.java:52)\n\tat io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:117)\n\tat io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:90)\n\tat io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)\n\tat org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)\n\tat org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)\n\tat org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)\n\tat org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\nCaused by: java.nio.charset.MalformedInputException: Input length = 1\n\tat java.base/java.nio.charset.CoderResult.throwException(CoderResult.java:274)\n\tat java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:820)\n\tat org.jboss.resteasy.util.Encode.decodeBytes(Encode.java:248)\n\t... 15 more"}
```

### How to Reproduce?

The simplest way to reproduce is is using the example project:
```
git clone https://github.com/quarkusio/quarkus-quickstarts.git
cd quarkus-quickstarts
./mvnw compile quarkus:dev
```
and then, in another terminal
```
curl -v http://localhost:8080/fruits/%FF
```

### Output of `uname -a` or `ver`

Darwin hostname.local 21.1.0 Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12) OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.5.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

_No response_

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/MalformedURLTestCase.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/MalformedURLTestCase.java
@@ -1,0 +1,31 @@
+package io.quarkus.resteasy.test;
+
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.Response;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MalformedURLTestCase {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(GreetingResource.class)
+                    .addAsResource(new StringAsset("greeting.message=hello"), "application.properties"));
+
+    @Test
+    public void testMalformedURL() {
+        // using JAX-RS client here because RestAssures seems to be doing some cleansing of the URL
+        Response response = ClientBuilder
+                .newClient().target("http://localhost:"
+                        + ConfigProvider.getConfig().getValue("quarkus.http.test-port", Integer.class) + "/%FF")
+                .request().get();
+        Assertions.assertEquals(response.getStatus(), 400);
+    }
+}

--- a/extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/VertxRequestHandler.java
+++ b/extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/VertxRequestHandler.java
@@ -3,6 +3,7 @@ package io.quarkus.resteasy.runtime.standalone;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.charset.MalformedInputException;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.Executor;
@@ -98,6 +99,21 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
     }
 
     private void dispatch(RoutingContext routingContext, InputStream is, VertxOutput output) {
+        ResteasyUriInfo uriInfo;
+        try {
+            uriInfo = VertxUtil.extractUriInfo(routingContext.request(), rootPath);
+        } catch (Exception e) {
+            if (e.getCause() instanceof MalformedInputException) {
+                log.debug(e.getCause());
+                routingContext.response().setStatusCode(400);
+            } else {
+                log.debug(e);
+                routingContext.response().setStatusCode(500);
+            }
+            routingContext.response().end();
+            return;
+        }
+
         ManagedContext requestContext = Arc.container().requestContext();
         requestContext.activate();
         routingContext.remove(QuarkusHttpUser.AUTH_FAILURE_HANDLER);
@@ -114,7 +130,7 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
         try {
             Context ctx = vertx.getOrCreateContext();
             HttpServerRequest request = routingContext.request();
-            ResteasyUriInfo uriInfo = VertxUtil.extractUriInfo(request, rootPath);
+
             ResteasyHttpHeaders headers = VertxUtil.extractHttpHeaders(request);
             HttpServerResponse response = request.response();
             VertxHttpResponse vertxResponse = new VertxHttpResponse(request, dispatcher.getProviderFactory(),

